KS-15-01-animals-move

DenisGubernatorov opened the pull request:
1) Реализованы интерфейсы Fly и Swim  с дефолтной реализацией методов doFle и doSwim  соответственно 2) Создан классы Carp(Карась) Seagull(Чайка) и Duck(Утка)  которые реализуют интерфейсы типа передвижения

DenisGubernatorov committed:
KS-15-01-animals-move
1) Реализованы интерфейсы Fly и Swim  с дефолтной реализацией методов doFle и doSwim  соответственно
2) Создан классы Carp(Карась) Seagull(Чайка) и Duck(Утка)  которые реализуют интерфейсы типа передвижения

ievetrovdev commented on src/main/kotlin/lesson15/lesson15_task1.kt:
```diff
@@ -1 +1,42 @@
 package lesson15
+
+interface Fly {
```
Хорошо, но нарушается принцип единой ответственности (см. [SOLID](https://habr.com/ru/companies/productivity_inside/articles/505430/)) потому что теперь я по ошибке или случайно могу заставить карася полететь и все сломается

ievetrovdev commented on src/main/kotlin/lesson15/lesson15_task1.kt:
```diff
@@ -1 +1,42 @@
+class Duck(
+    override val name: String,
+) : Swim,
+    Fly
```
Здесь можно в одну строку)

DenisGubernatorov replied:
> Здесь можно в одну строку)
А у меня klint подчеркивает. И при коммите стоит реформат кода
![image](https://github.com/user-attachments/assets/550670cc-4917-415e-bcb2-72efc091c6b1)
![image](https://github.com/user-attachments/assets/c4974590-3193-446b-a07f-da8e0e4d4940)

DenisGubernatorov committed:
KS-15-01-animals-move
1) Убрал из интерфейса поле name. Что бы не было возможности подменить название и не получилось "Карп летит"
2) Но добавил в классах поля с русским названием классов. Для возможности красивого вывода.

ievetrovdev reviewed (changes requested):
Классы `Carp`, `Seagull`, и `Duck` реализуют интерфейсы, но не реализуют методы этих интерфейсов (то есть методы `doFly` и `doSwim` не переопределяются явно). Несмотря на то, что в Kotlin интерфейсы могут иметь методы с реализацией, явное переопределение методов в классах помогает избежать ошибок и улучшает читаемость кода.

DenisGubernatorov committed:
KS-15-01-animals-move
1) для интерфейсов убрана дефолтная реализация. Методы должны быть переопределены в каждом классе, имплементирующем интерфейс

DenisGubernatorov commented:
> Классы `Carp`, `Seagull`, и `Duck` реализуют интерфейсы, но не реализуют методы этих интерфейсов (то есть методы `doFly` и `doSwim` не переопределяются явно). Несмотря на то, что в Kotlin интерфейсы могут иметь методы с реализацией, явное переопределение методов в классах помогает избежать ошибок и улучшает читаемость кода.

Ну да. Я как бы и учитывал момент, что в интерфейсах может быть дефолтная реализация метода

ievetrovdev commented:
> > Классы `Carp`, `Seagull`, и `Duck` реализуют интерфейсы, но не реализуют методы этих интерфейсов (то есть методы `doFly` и `doSwim` не переопределяются явно). Несмотря на то, что в Kotlin интерфейсы могут иметь методы с реализацией, явное переопределение методов в классах помогает избежать ошибок и улучшает читаемость кода.
> 
> Ну да. Я как бы и учитывал момент, что в интерфейсах может быть дефолтная реализация метода

Это не совсем корректная реализация поведения по умолчанию. 
Все же интерфейсы предназначены для определения поведения, а не для хранения состояния, а определяя свойство name в интерфейсе, любой класс, реализующий этот интерфейс, зависит от этого свойства. Но что, если классу не нужно иметь это свойство, а нужно только выполнять действие, описанное в интерфейсе?

ievetrovdev reviewed (approved):
Теперь все корректно, апрув!
